File: ReverseLinkedList/ReverseLinkedList.cpp
// ReverseLinkedList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

ListNode* ReverseNodes(ListNode* pNode, ListNode** pNewHead)
{
    if (pNode->next == nullptr)
    {
        *pNewHead = pNode;
        return pNode;
    }
    else
        ReverseNodes(pNode->next, pNewHead)->next = pNode;
    return pNode;
}

ListNode* ReverseLinkedList(ListNode* pHead)
{
    ListNode* pNewHead = nullptr;

    ListNode* pNode = ReverseNodes(pHead, &pNewHead);
    pNode->next = nullptr;
    return pNewHead;
}

std::ostream& operator <<(std::ostream& st, const ListNode* pNode)
{
    st << pNode->data;
    if (pNode->next != nullptr)
    {
        st << " ," << pNode->next;
    }
    else
        st << std::endl;

    return st;
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of nodes in list: ";
        std::cin >> count;
        if (count == 0)
            break;

        ListNode* pOrignalHead = nullptr;
        ListNode* pOriginalTail = GetListFromInput(&pOrignalHead, count);

        std::cout << "Original: ";
        std::cout << pOrignalHead;

        ListNode* pReversed = ReverseLinkedList(pOrignalHead);

        std::cout << "Reversed: ";
        std::cout << pReversed;
        FreeList(pReversed);
    }
}

File: Go/ReverseLinkedLIst.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) *ListNode {
	if len(v) == 0 {
		return nil
	}

	if len(v) == 1 {
		return &ListNode{v[0], nil}
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead
}

func ReverseList(pNode *ListNode) (*ListNode, *ListNode) {
	if pNode.next == nil {
		return pNode, pNode
	}

	pHead, pNewNode := ReverseList(pNode.next)
	pNewNode.next = pNode
	return pHead, pNode
}

func ReverseLinkedList(head *ListNode) *ListNode {
	pHead, _ := ReverseList(head)
	head.next = nil
	return pHead
}

func (pNode *ListNode) Print() {
	fmt.Print(pNode.data)
	if pNode.next != nil {
		fmt.Print(", ")
		pNode.next.Print()
	} else {
		fmt.Println("")
	}
}

func main() {
	for {
		fmt.Print("Number of elements: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := GetArrayFromInput(count, "Please enter the list: ")
		pHead := GetListFromArray(vec)

		pHead = ReverseLinkedList(pHead)
		pHead.Print()
	}
}
